### PR TITLE
[mono][jit] read DebuggableAttribute(DebuggingModes) ctor; move to metadata/

### PR DESCRIPTION
**Pull request by lambdageek**
Pull `is_jit_optimizer_disabled` out of mini into metadata.

Also support reading the `DebuggingModes` ctor

---

This is needed to properly support https://github.com/dotnet/runtime/issues/47274 (the `DOTNET_MODIFIABLE_ASSEMBLIES=debug` env var for hot reload)

### CODE REVIEW

**Comment by lambdageek** [comment]
@vargaz Looks like Roslyn switched at some point from emitting `DebuggableAttribute(boolean, boolean)` to `DebuggabelAttribute(DebuggingMode)` and we've been silently allowing jit optimizations even when we shouldn't have been.

File: src/mono/mono/metadata/assembly.c [hunk 1]
```diff
@@ -3720,3 +3722,78 @@ mono_asmctx_get_name (const MonoAssemblyContext *asmctx)
...
+gboolean
+mono_assembly_is_jit_optimizer_disabled (MonoAssembly *ass)
+{
+	ERROR_DECL (error);
+
+	g_assert (ass);
+	if (ass->jit_optimizer_disabled_inited)
+		return ass->jit_optimizer_disabled;
+
+	MonoClass *klass =mono_class_try_get_debuggable_attribute_class ();
```
Comment (thaystg) [inline src/mono/mono/metadata/assembly.c:3744]:
```suggestion
	MonoClass *klass = mono_class_try_get_debuggable_attribute_class ();
```

File: src/mono/mono/metadata/assembly.c [hunk 2]
```diff
@@ -3720,3 +3722,78 @@ mono_asmctx_get_name (const MonoAssemblyContext *asmctx)
...
+	if (!klass) {
+		/* Linked away */
+		ass->jit_optimizer_disabled = FALSE;
+		mono_memory_barrier ();
+		ass->jit_optimizer_disabled_inited = TRUE;
+		return FALSE;
+	}
+
+	gboolean disable_opts = FALSE;
+	MonoCustomAttrInfo* attrs =mono_custom_attrs_from_assembly_checked (ass, FALSE, error);
```
Comment (thaystg) [inline src/mono/mono/metadata/assembly.c:3755]:
```suggestion
	MonoCustomAttrInfo* attrs = mono_custom_attrs_from_assembly_checked (ass, FALSE, error);
```

**Review by thaystg — APPROVED** [review]
LGTM.

**Comment by lambdageek** [comment]
@vargaz  can you remember the details of https://github.com/dotnet/runtime/commit/1d0195c57057cf8033f28c0d901e37022c149957 (I can't see Xamarin bug 604486) 

It doesn't look like `mcs`, Roslyn or F# ever emitted the `DebuggableAttribute(boolean,boolean)` constructor custom attribute.  Always the `DebuggerModes` one.

I'm wondering if enabling this in mono's JIT is a good idea now.  Or if we should continue ignoring it as apparently it hasn't done any harm in the last 7 years.

**Pull request merged**